### PR TITLE
GCC: ARM_CM0: Fix L6286E error on Keil MDK

### PR DESCRIPTION
**Pull request by TomasGalbickaNXP**
Keil MDK Compilation for CM0 core reports L6286E: Relocation #REL:0 in portasm.o

Description
-----------
Change the .b instruction to .bl with higher range to solve error reported by MDK descibed bellow.

Fix:
Error: L6286E: Relocation #REL:0 in portasm.o(.text.SVC_Handler) with respect to vPortSVCHandler_C. Value(0x1a04) out of range(-0x800 - 0x7fe) for (R_ARM_THM_JUMP11)

Compiler: Keil MDK ARMClang 6.22.0

https://developer.arm.com/documentation/ka002847/latest/ https://developer.arm.com/documentation/dui0496/m/Linker-Errors-and-Warnings/List-of-the-armlink-error-and-warning-messages

See the issue #1130 

Test Steps
-----------
See the issue #1130 

To Reproduce
git clone https://github.com/TomasGalbickaNXP/freertos-port-cm0-core-issue.git
Checkout the Example with added CM0 MPU port.
git checkout 40af92d06408450e11ca10ad17cd578760fd861a
Open MDK Project and build. It will fail with error reported above.
Checkout fix
git checkout 72a9d6246277ae32594bab69fef66c6e13af3cd1
Open MDK Project and build.
The build reports no Errors.

Checklist:
----------
- [ x ] I have tested my changes. No regression in existing tests.
- [    ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------
#1130 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: portable/GCC/ARM_CM0/portasm.c [hunk 1]
```diff
@@ -456,11 +456,11 @@ void vClearInterruptMask( __attribute__( ( unused ) ) uint32_t ulMask ) /* __att
             "                                \n"
             " stack_on_psp:                  \n"
             "     mrs r0, psp                \n"
-            "     b route_svc                \n"
+            "     bl route_svc               \n"
```
Comment (urutva) [inline portable/GCC/ARM_CM0/portasm.c:459]: From [Armv6-M architecture manual](https://developer.arm.com/documentation/ddi0419/latest/), section A4.3 Branch instructions, the range supported by `B` instruction is +/- 2KB which is not sufficient to jump to `vPortSVCHandler_C` in the [issue](https://github.com/FreeRTOS/FreeRTOS-Kernel/issues/1130) you have raised. Changing from instruction `B` to `BL` solves this issue. Please note that `B` is a 16-bit thumb instruction, where as `BL` is 32-bit thumb instruction. The `route_svc` label is is in the same function and is just after couple of instructions and it should still be reachable with `B` instruction and save 2 bytes of code, unless I'm missing something. Same suggestion for the other change involving `route_svc`.

**Review by urutva — CHANGES_REQUESTED** [review]

**Comment by aggarg** [comment]
Can you share the project that we can use to repro the problem?

File: portable/GCC/ARM_CM0/portasm.c (hunk 1 above)
Comment (TomasGalbickaNXP) [reply portable/GCC/ARM_CM0/portasm.c:459]: Hi, Okay for the `route_svc` jump the `B` should be sufficient in this case. I will update.
Thanks!

**Comment by TomasGalbickaNXP** [comment]
Hi @aggarg Please refer to the Issue there is To Reproduce with the link to git repo with the example project. I will also update this PR.
https://github.com/FreeRTOS/FreeRTOS-Kernel/issues/1130

**Comment by aggarg** [comment]
> Hi @aggarg Please refer to the Issue there is To Reproduce with the link to git repo with the example project. I will also update this PR. #1130

Thank you. I will give it a try and get back.

**Comment by aggarg** [comment]
Branch and Link (bl) instruction changes the value of Link Register (LR) which needs to be preserved. There are 2 ways to solve this problem -
1. Push and pop the Link Register (LR) in the way we do for `system_call_enter` and `system_call_exit`:
    ```
    push {lr}
    bl vPortSVCHandler_C
    pop {pc}
    ```
    This has a drawback - the system call to start the scheduler (`portSVC_START_SCHEDULER`) never returns and therefore, the corresponding pop is never executed. As a result, we end up wasting 4 bytes on MSP.

2. Alternative is to use Branch and Exchange (BX) instruction which does not have the range restriction. The following patch implements this solution -  [use_bx.patch](https://github.com/user-attachments/files/16782642/use_bx.patch). I have tested this solution on a Cortex-M0+ platform and I have checked that your project builds too with this change. I suggest that we go with this solution. If you agree, would you please apply this patch in this PR.

Thanks.

**Comment by TomasGalbickaNXP** [comment]
Great! Thanks for your fix.
Yes agreed `bx` instruction will be better here.

I have applied you provided patch.

File: portable/GCC/ARM_CM0/portasm.c [hunk 2]
```diff
@@ -443,49 +443,52 @@ void vClearInterruptMask( __attribute__( ( unused ) ) uint32_t ulMask ) /* __att
...
+            "     ldr r3, [r0, #24]                                 \n"
+            "     subs r3, #2                                       \n"
+            "     ldrb r2, [r3, #0]                                 \n"
+            "     ldr r3, =%0                                       \n"
+            "     cmp r2, r3                                        \n"
+            "     blt system_call_enter                             \n"
+            "     ldr r3, =%1                                       \n"
+            "     cmp r2, r3                                        \n"
+            "     beq system_call_exit                              \n"
+            "     ldr r3, vPortSVCHandler_C_Address                 \n"
```
Comment (urutva) [inline portable/GCC/ARM_CM0/portasm.c:475]:
```suggestion
            "     ldr r3, = vPortSVCHandler_C                 \n"
```
@TomasGalbickaNXP @aggarg Apologies for the nit. With this style, we can avoid the constant pool defined at the end of the function and let the linker handle the needed alignement.

**Review by urutva — CHANGES_REQUESTED** [review]

**Comment by aggarg** [comment]
Thanks @urutva for pointing that out. Here is the updated patch - [use_bx_updated.patch](https://github.com/user-attachments/files/16785695/use_bx_updated.patch). I have tested it on M0 for both MPU enable and disable. 

@TomasGalbickaNXP please use this patch instead. Apologies for back and forth.

**Comment by TomasGalbickaNXP** [comment]
OK patch applied.

**Review by urutva — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Comment by aggarg** [comment]
Thank you for your contribution!

**Comment by TomasGalbickaNXP** [comment]
Thank you for great assistance!

**Pull request merged**